File: algorithms/math/miller-rabin/java/MillerRabin.java
public class MillerRabin {

    public static int millerRabin(int n) {
        if (n < 2) return 0;
        if (n < 4) return 1;
        if (n % 2 == 0) return 0;

        int r = 0;
        long d = n - 1;
        while (d % 2 == 0) {
            r++;
            d /= 2;
        }

        int[] witnesses = {2, 3, 5, 7};
        for (int a : witnesses) {
            if (a >= n) continue;

            long x = modPow(a, d, n);
            if (x == 1 || x == n - 1) continue;

            boolean found = false;
            for (int i = 0; i < r - 1; i++) {
                x = modPow(x, 2, n);
                if (x == n - 1) {
                    found = true;
                    break;
                }
            }

            if (!found) return 0;
        }

        return 1;
    }

    private static long modPow(long base, long exp, long mod) {
        long result = 1;
        base %= mod;
        while (exp > 0) {
            if (exp % 2 == 1) result = result * base % mod;
            exp /= 2;
            base = base * base % mod;
        }
        return result;
    }
}

File: algorithms/math/miller-rabin/swift/MillerRabin.swift
func millerRabin(_ n: Int) -> Int {
    if n < 2 { return 0 }
    if n < 4 { return 1 }
    if n % 2 == 0 { return 0 }

    func modPow(_ base: Int, _ exp: Int, _ mod: Int) -> Int {
        var result = 1
        var b = base % mod
        var e = exp
        while e > 0 {
            if e % 2 == 1 { result = result * b % mod }
            e /= 2
            b = b * b % mod
        }
        return result
    }

    var r = 0
    var d = n - 1
    while d % 2 == 0 { r += 1; d /= 2 }

    let witnesses = [2, 3, 5, 7]
    for a in witnesses {
        if a >= n { continue }

        var x = modPow(a, d, n)
        if x == 1 || x == n - 1 { continue }

        var found = false
        for _ in 0..<(r - 1) {
            x = modPow(x, 2, n)
            if x == n - 1 { found = true; break }
        }

        if !found { return 0 }
    }

    return 1
}
